### ISSUE TITLE
E0733 (async recursion) does not detected for functions calling each other

### ISSUE THREAD

**Issue by neonaot** [intellij-rust/intellij-rust#9319]
## Environment

* **IntelliJ Rust plugin version:** 0.4.178.4857-222-nightly
* **Rust toolchain version:** 1.59.0-nightly (cfa3fe5af 2021-12-31) x86_64-apple-darwin
* **IDE name and version:** CLion 2022.2.1 (CL-222.3739.54)
* **Operating system:** macOS 12.5.1
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.emulate.terminal, org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc

## Problem description
<img width="781" alt="image" src="https://user-images.githubusercontent.com/49211026/188612909-79aa40f6-8a04-4eca-9e83-532c4437dafe.png">
Note that it is enough to add `![async_recursion]` to any of the functions.

Expected:
<img width="572" alt="image" src="https://user-images.githubusercontent.com/49211026/188613694-f802b258-614e-4a01-82bf-2cc563b82501.png">

## Steps to reproduce
```rust
//#[async_recursion]
async fn foo() {
    bar().await;
}

async fn bar() {
    foo().await;
}
```

**Comment by Undin** [comment]
Just want to mention that supporting this feature requires interprocedural analysis => it's a quite hard task for IDE since you need to analyze the whole program in the worst case on each change
